### DISCUSSION TITLE
增加自定义内容选项，支持用大语言模型翻译后，对翻译结果与自定义内容比较，如果相等，则将翻译结果替换为空

### DISCUSSION THREAD

**Discussion by zhengshuo123** [immersive-translate/immersive-translate#2140] [Ideas]
### 新功能 | New Feature

增加自定义内容选项，支持用大语言模型翻译后，对翻译结果与自定义内容比较，如果相等，则将翻译结果替换为空

### 为什么您想要添加这些功能 | Why you want to add these features

翻译电子书总是翻译代码和命令之类的内容。即使修改提示词，让他不要翻译，也总是会输出一些内容或者原封不动重复文本。
感觉大模型对于不输出的提示词不敏感，所以想了一个办法，通过提示词告诉他将代码或命令翻译成特定的内容，比如NULL，然后插件再来将它返回的结果，这个例子是NULL，与自定义内容比较，如果一样，就翻译输出是空，就可以方便的通过提示词解决翻译代码和命令的问题

**Comment by theowenyoung** [comment]
目前有个比较hack 的做法，你可以让大模型输出以下词语，沉浸式翻译就会替换为空：

```
      "ignoreResRegexs": [
        "^抱歉.*要求",
        "^抱歉.*请求",
        "^抱歉.*翻译",
        "^很抱歉.*翻译",
        "^我很抱歉.*翻译",
        "^对不起，我无法直接翻译",
        "^抱歉.*我无法",
        "^I'm sorry, but I cannot",
        "^I'm sorry, but I cannot provide",
        "^I'm sorry, I can't assist with that.",
        "^这句话的内容不适合在此平台上讨论",
        "^这句话不适合在公共场合讨论",
        "地道的翻译引擎，你只返回译文，不含任何解释"
      ],
```

比如你让大模型返回：“对不起，我无法直接翻译”

**Reply by zhengshuo123** [reply]
> 目前有个比较hack 的做法，你可以让大模型输出以下词语，沉浸式翻译就会替换为空：
> 
> ```
>       "ignoreResRegexs": [
>         "^抱歉.*要求",
>         "^抱歉.*请求",
>         "^抱歉.*翻译",
>         "^很抱歉.*翻译",
>         "^我很抱歉.*翻译",
>         "^对不起，我无法直接翻译",
>         "^抱歉.*我无法",
>         "^I'm sorry, but I cannot",
>         "^I'm sorry, but I cannot provide",
>         "^I'm sorry, I can't assist with that.",
>         "^这句话的内容不适合在此平台上讨论",
>         "^这句话不适合在公共场合讨论",
>         "地道的翻译引擎，你只返回译文，不含任何解释"
>       ],
> ```
> 
> 比如你让大模型返回：“对不起，我无法直接翻译”

OK，我试试

**Reply by zhengshuo123** [reply]
> 目前有个比较hack 的做法，你可以让大模型输出以下词语，沉浸式翻译就会替换为空：
> 
> ```
>       "ignoreResRegexs": [
>         "^抱歉.*要求",
>         "^抱歉.*请求",
>         "^抱歉.*翻译",
>         "^很抱歉.*翻译",
>         "^我很抱歉.*翻译",
>         "^对不起，我无法直接翻译",
>         "^抱歉.*我无法",
>         "^I'm sorry, but I cannot",
>         "^I'm sorry, but I cannot provide",
>         "^I'm sorry, I can't assist with that.",
>         "^这句话的内容不适合在此平台上讨论",
>         "^这句话不适合在公共场合讨论",
>         "地道的翻译引擎，你只返回译文，不含任何解释"
>       ],
> ```
> 
> 比如你让大模型返回：“对不起，我无法直接翻译”

![image](https://github.com/user-attachments/assets/9e7d8912-2271-427c-a115-0ffc24de1ef7)

好像不行

**Reply by theowenyoung** [reply]
似乎是 逗号 是yk w英文逗号？

![CleanShot 2024-09-27 at 11 02 56@2x](https://github.com/user-attachments/assets/768d91b2-2374-4ddb-a5be-0e9338f8d1c4)

可以测试下其他的？

比如

`I'm sorry, but I cannot`

`这句话的内容不适合在此平台上讨论`

**Reply by zhengshuo123** [reply]
> 似乎是 逗号 是yk w英文逗号？
> 
> ![CleanShot 2024-09-27 at 11 02 56@2x](https://private-user-images.githubusercontent.com/62473795/371376934-768d91b2-2374-4ddb-a5be-0e9338f8d1c4.png?jwt=eyJhbGciOiJIUzI1NiIsInR5cCI6IkpXVCJ9.eyJpc3MiOiJnaXRodWIuY29tIiwiYXVkIjoicmF3LmdpdGh1YnVzZXJjb250ZW50LmNvbSIsImtleSI6ImtleTUiLCJleHAiOjE3Mjc0MTIxOTEsIm5iZiI6MTcyNzQxMTg5MSwicGF0aCI6Ii82MjQ3Mzc5NS8zNzEzNzY5MzQtNzY4ZDkxYjItMjM3NC00ZGRiLWE1YmUtMGU5MzM4ZjhkMWM0LnBuZz9YLUFtei1BbGdvcml0aG09QVdTNC1ITUFDLVNIQTI1NiZYLUFtei1DcmVkZW50aWFsPUFLSUFWQ09EWUxTQTUzUFFLNFpBJTJGMjAyNDA5MjclMkZ1cy1lYXN0LTElMkZzMyUyRmF3czRfcmVxdWVzdCZYLUFtei1EYXRlPTIwMjQwOTI3VDA0MzgxMVomWC1BbXotRXhwaXJlcz0zMDAmWC1BbXotU2lnbmF0dXJlPWE1ZTdjMmY1OWFiZDhmN2FiM2RkYzBhMzMwZDgyMjIzMGU0ZTkwNWU4YjI4OTRkYzZlZTYyNjUzNDg1MzQyYzcmWC1BbXotU2lnbmVkSGVhZGVycz1ob3N0In0.iRex65drRdRecgIku7jMtVdyAAFUt4Wo6Cbo5RecjQk)
> 
> 可以测试下其他的？
> 
> 比如
> 
> `I'm sorry, but I cannot`
> 
> `这句话的内容不适合在此平台上讨论`

![image](https://github.com/user-attachments/assets/64b75b70-ddf4-45d2-8e1d-d42d4ec22b0e)
![image](https://github.com/user-attachments/assets/f5121534-88ea-4418-9922-daefbaad0550)
我通过这种方式测试了几个，在输入框空格三下让它调用大模型，都返回了，没有替换

**Reply by theowenyoung** [reply]
在输入框里翻译时，如果遇到这类返回，正常情况下，是不会翻译。在网页阅读中的行为，也是如此，这种响应的行为是不翻译，我测试的prompt：
![CleanShot 2024-09-27 at 14 20 57@2x](https://github.com/user-attachments/assets/c111b27a-79e8-4c9f-8f2d-5506ba1a3894)

https://github.com/user-attachments/assets/dd5bc450-e3c6-4b92-92d1-fd18fa9b8f96

**Reply by zhengshuo123** [reply]
https://github.com/user-attachments/assets/2f2663fe-1c42-456c-ba03-a2de1fc42595

我用的是edge浏览器，
![image](https://github.com/user-attachments/assets/f2631c20-70f5-4ff3-9814-9d1a61b7d75e)
siliconflow提供的免费版大模型测试的，这个和模型应该是没关系的吧，版本也是最新的，有点奇怪

**Reply by zhengshuo123** [reply]
https://github.com/bookfere/Ebook-Translator-Calibre-Plugin/tree/master
换成了calibre插件，开源的，自己修改代码后，解决了翻译代码的问题，得配合calibre使用，也支持自定义大模型接入，更稳定
